### PR TITLE
Feat/add support for bitbucket sync

### PR DESCRIPTION
**Pull request by TheeMattOliver**
## tl;dr

Draft PR to assess the state of Bitbucket sync and token storage for Figma Tokens. Goal: bring Bitbucket sync for Figma Tokens up to speed with other git providers

## What's working

- authenticate and check correct permissions
- push custom commit message
- push commit data content as new commit to main branch
- push commit data content as new commit to designated branch
- creating pull requests

## What's left to do 

- pull data from Bitbucket repo by implementing the `read` method in `BitbucketTokenStorage.ts`
- create a tree structure from the contents of a Bitbucket directory to support multifile sync
- handle paginated lists of branches in `fetchBranches` in `BitbucketTokenStorage.ts`
- improve unit test coverage

## Files of interest:

remaining work to be done in this file &rarr; `src/storage/BitbucketTokenStorage.ts`

And for reference: 

`src/storage/RemoteTokenStorage.ts`
`src/app/store/remoteTokens.tsx`
`src/app/store/updateSources.tsx`
`src/store/bitbucket/bitbucket.tsx`

`src/app/components/SyncSettings.tsx`
`src/app/components/StorageItemForm/BitbucketForm`
`src/app/components/PushDialog`

## Note on the `read` method
Other git providers have wrapper methods that have provided a `createTree` method to create a tree structure from the contents of a directory. See [here](https://octokit.github.io/rest.js/v18#git-create-tree) and [here](https://docs.gitlab.com/ee/api/repositories.html), for example.

Neither the [batteries-included wrapper library](https://github.com/MunifTanjim/node-bitbucket) we're using nor the [raw Bitbucket API](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get) provide that.

I've tried to look for guidance in the Atlassian docs but all I can find is that [supposedly this endpoint supports recursion if you specify additionally the max_depth parameter and indicate how many levels you want to have.](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get) Haven't been able to try that. My feeling is that we may have to craft a recursive traversal method similar to `git ls-tree` to support `read` in `BitbucketTokenStorage`. I could be wrong about that and I am open to suggestions!

## Note on unit tests 
The methods that are implemented are working in the plug-in, but I think I'm doing something incorrectly in the way I'm mocking those functions or passing return values into the mocked functions in Jest. Open to suggestions for correctly mocking them to improve test coverage.

Those are in `/storage/___tests___/BitbucketTokenStorage.test.ts`.

### CODE REVIEW

**Comment by LiamMartens** [comment]
The parts that are already implemented look pretty solid. If we want to put the UI part of this behind a feature flag all we need to do is the following:
- Add the flag variable in `flags.d.ts` just for the typings. We can add something like `bitbucketSync`
- In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately

CC @six7

**Comment by six7** [comment]
> The parts that are already implemented look pretty solid. If we want to put the UI part of this behind a feature flag all we need to do is the following:
> 
> * Add the flag variable in `flags.d.ts` just for the typings. We can add something like `bitbucketSync`
> * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately
> 
> CC @six7

Sounds good. Let's do that. 

@TheeMattOliver Can you make sure to add those flags? Then we could merge this first version and later on expand on that. It's in a state where it's ready to be tested though, right? I think we'll have a large amount of testers for this as it's such a highly requested feature.

I've setup a `bitBucketSync` flag in Launch Darkly and enabled it for your user. Once you implemented it you can test it by removing your key then the flag should be false.

**Comment by TheeMattOliver** [comment]
Thanks!

 > * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately

I'm not sure I understand how to use this since I've never actually used feature flags myself 🙈 If you point me in the right direction I can read up on it though!

> @TheeMattOliver Can you make sure to add those flags types? It's called `bitBucketSync` now. 

Done, no problem

> It's in a state where it's ready to be tested though, right? I think we'll have a large amount of testers for this as it's such a highly requested feature.

Before you let people start mashing on it I think we should probably implement the `read` method

> I've setup a `bitBucketSync` flag in Launch Darkly and enabled it for your user. Once you implemented it you can test it by removing your key then the flag should be false.

**Comment by six7** [comment]
> Thanks!
> 
> > * In the `SyncSettings.tsx` file you can then use our wrapped `useFlags` hook (from `app/components/LaunchDarkly`) to check for the availability of the flag and hide/show the UI to select the new provider accordingly. Then we could be ok to merge this in (after some of the comments) and pick up the finalization of the new provider separately
> 
> I'm not sure I understand how to use this since I've never actually used feature flags myself 🙈 If you point me in the right direction I can read up on it though!

For example in `ThemeSelector.tsx` we check the presence of the `tokenThemes` flag. If true, show something, if not, don't.

It would be like this for the bitBucketSync flag:
```
import { useFlags } from '../LaunchDarkly';
const { bitBucketSync } = useFlags();

…

if (bitBucketSync) {
  // code when flag is true
}

```

**Comment by TheeMattOliver** [comment]
Got it, thanks! Please hold, will update and push

File: src/utils/deepMerge.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/* ts-deep-merge, unlicensed, https://github.com/voodoocreation/ts-deepmerge/blob/master/src/index.ts */
```
Comment (six7) [inline src/utils/deepMerge.ts:1]: Do we need this file? It's not used anywhere and it's unlicensed?

**Review by six7 — CHANGES_REQUESTED** [review]
Can we remove the deepMerge file? Other than that I'd be happy to merge this in 👍

**Comment by TheeMattOliver** [comment]
phew the commit history is a mess but this branch should be good to merge in now ✅

File: src/storage/BitbucketTokenStorage.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
+import { Bitbucket } from 'bitbucket';
+import { RemoteTokenStorageFile } from './RemoteTokenStorage';
+import {
+  GitMultiFileObject, GitSingleFileObject, GitStorageMetadata, GitTokenStorage,
+} from './GitTokenStorage';
```
Comment (six7) [inline src/storage/BitbucketTokenStorage.ts:5]:
```suggestion
import { GitStorageMetadata, GitTokenStorage } from './GitTokenStorage';
```

File: src/storage/BitbucketTokenStorage.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  // https://developer.atlassian.com/cloud/bitbucket/rest/api-group-refs/#api-repositories-workspace-repo-slug-refs-branches-post
+  public async createBranch(branch: string, source?: string) {
+    try {
+      const originRef = `refs/${source || this.branch}`;
+      const newRef = `refs/${branch}`;
+      const originBranch = await this.bitbucketClient.listRefs({
+        workspace: this.owner,
+        repo_slug: this.repository,
+        ref: originRef,
+      });
```
Comment (six7) [inline src/storage/BitbucketTokenStorage.ts:63]:
```suggestion
      });
      console.log('originBranch', originBranch);
      console.log('newRef', newRef);
```

File: src/storage/BitbucketTokenStorage.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  // https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get
+  // Equivalent to directly hitting /2.0/repositories/{username}/{repo_slug}/src/{commit}/{path} without having to know the name or SHA1 of the repo's main branch.
+  public async read(): Promise<RemoteTokenStorageFile<GitStorageMetadata>[]> {
+    try {
+      const response = await this.bitbucketClient.repositories.get({
+        workspace: this.owner,
+        repo_slug: this.repository,
+        // path: this.path,
+        // ref: this.branch,
+      });
```
Comment (six7) [inline src/storage/BitbucketTokenStorage.ts:111]:
```suggestion
      });
      console.log('response', response);
```

File: src/app/hooks/useIsGitMultiFileEnabled.ts [hunk 5]
```diff
@@ -6,10 +6,15 @@ import { StorageProviderType } from '@/constants/StorageProviderType';
...
+    () => Boolean(
+      multiFileSync
+          && (api?.provider === StorageProviderType.GITHUB
+            || api?.provider === StorageProviderType.GITLAB
+            || api?.provider === StorageProviderType.ADO
+            || api?.provider === StorageProviderType.BITBUCKET)
+          && !api?.filePath?.endsWith('.json'),
+    ),
+    [api, multiFileSync],
+  );
```
Comment (six7) [inline src/app/hooks/useIsGitMultiFileEnabled.ts:19]:
```suggestion
  return useMemo(
    () => Boolean((api?.provider === StorageProviderType.GITHUB
            || api?.provider === StorageProviderType.GITLAB
            || api?.provider === StorageProviderType.ADO
            || api?.provider === StorageProviderType.BITBUCKET)
          && !api?.filePath?.endsWith('.json')),
    [api],
  );
```

File: src/storage/__tests__/BitbucketTokenStorage.test.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
+import { TokenTypes } from '@/constants/TokenTypes';
+import { TokenSetStatus } from '@/constants/TokenSetStatus';
```
Comment (six7) [inline src/storage/__tests__/BitbucketTokenStorage.test.ts:2]:
```suggestion
// import { TokenTypes } from '@/constants/TokenTypes';
// import { TokenSetStatus } from '@/constants/TokenSetStatus';
```

File: src/storage/__tests__/BitbucketTokenStorage.test.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        ],
+      },
+    }));
+
+    expect(await storageProvider.canWrite()).toBe(false);
+  });
+
+  it('canWrite should return true if user has admin or write permissions', async () => {
+    mockGetAuthedUser.mockImplementationOnce(() => Promise.resolve({
+      currentUser: { data: { account_id: '' } },
```
Comment (six7) [inline src/storage/__tests__/BitbucketTokenStorage.test.ts:51]:
```suggestion
      data: { account_id: '123' },
```

File: src/storage/__tests__/BitbucketTokenStorage.test.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+                },
+              },
+              null,
+              2,
+            ),
+          },
+        },
+      ],
+    });
+  });
```
Comment (six7) [inline src/storage/__tests__/BitbucketTokenStorage.test.ts:174]:
```suggestion
  // it('should be able to write', async () => {
  //   mockListBranches.mockImplementationOnce(() => Promise.resolve({
  //     data: { values: [{ name: 'main' }] },
  //   }));

  //   mockCreateOrUpdateFiles.mockImplementationOnce(() => Promise.resolve({
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     branch: 'main',
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json':
  //               '{\n'
  //               + '  "$themes": [\n'
  //               + '    {\n'
  //               + '      "id": "light",\n'
  //               + '      "name": "Light",\n'
  //               + '      "selectedTokenSets": {\n'
  //               + '        "global": "enabled"\n'
  //               + '      }\n'
  //               + '    }\n'
  //               + '  ],\n'
  //               + '  "global": {\n'
  //               + '    "red": {\n'
  //               + '      "type": "color",\n'
  //               + '      "name": "red",\n'
  //               + '      "value": "#ff0000"\n'
  //               + '    }\n'
  //               + '  }\n'
  //               + '}',
  //         },
  //       },
  //     ],
  //   }));

  //   storageProvider.changePath('data/tokens.json');
  //   await storageProvider.write([
  //     {
  //       type: 'metadata',
  //       path: '$metadata.json',
  //       data: {},
  //     },
  //     {
  //       type: 'themes',
  //       path: '$themes.json',
  //       data: [
  //         {
  //           id: 'light',
  //           name: 'Light',
  //           selectedTokenSets: {
  //             global: TokenSetStatus.ENABLED,
  //           },
  //         },
  //       ],
  //     },
  //     {
  //       type: 'tokenSet',
  //       name: 'global',
  //       path: 'global.json',
  //       data: {
  //         red: {
  //           type: TokenTypes.COLOR,
  //           name: 'red',
  //           value: '#ff0000',
  //         },
  //       },
  //     },
  //   ], {
  //     commitMessage: '',
  //   });

  //   expect(mockCreateOrUpdateFiles).toBeCalledWith({
  //     branch: 'main',
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     createBranch: false,
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json': JSON.stringify(
  //             {
  //               $themes: [
  //                 {
  //                   id: 'light',
  //                   name: 'Light',
  //                   selectedTokenSets: {
  //                     global: TokenSetStatus.ENABLED,
  //                   },
  //                 },
  //               ],
  //               global: {
  //                 red: {
  //                   type: TokenTypes.COLOR,
  //                   name: 'red',
  //                   value: '#ff0000',
  //                 },
  //               },
  //             },
  //             null,
  //             2,
  //           ),
  //         },
  //       },
  //     ],
  //   });
  // });
```

Comment (six7) [inline src/storage/__tests__/BitbucketTokenStorage.test.ts:174]:
```suggestion
  // it('should be able to write', async () => {
  //   mockListBranches.mockImplementationOnce(() => Promise.resolve({
  //     data: { values: [{ name: 'main' }] },
  //   }));

  //   mockCreateOrUpdateFiles.mockImplementationOnce(() => Promise.resolve({
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     branch: 'main',
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json':
  //               '{\n'
  //               + '  "$themes": [\n'
  //               + '    {\n'
  //               + '      "id": "light",\n'
  //               + '      "name": "Light",\n'
  //               + '      "selectedTokenSets": {\n'
  //               + '        "global": "enabled"\n'
  //               + '      }\n'
  //               + '    }\n'
  //               + '  ],\n'
  //               + '  "global": {\n'
  //               + '    "red": {\n'
  //               + '      "type": "color",\n'
  //               + '      "name": "red",\n'
  //               + '      "value": "#ff0000"\n'
  //               + '    }\n'
  //               + '  }\n'
  //               + '}',
  //         },
  //       },
  //     ],
  //   }));

  //   storageProvider.changePath('data/tokens.json');
  //   await storageProvider.write([
  //     {
  //       type: 'metadata',
  //       path: '$metadata.json',
  //       data: {},
  //     },
  //     {
  //       type: 'themes',
  //       path: '$themes.json',
  //       data: [
  //         {
  //           id: 'light',
  //           name: 'Light',
  //           selectedTokenSets: {
  //             global: TokenSetStatus.ENABLED,
  //           },
  //         },
  //       ],
  //     },
  //     {
  //       type: 'tokenSet',
  //       name: 'global',
  //       path: 'global.json',
  //       data: {
  //         red: {
  //           type: TokenTypes.COLOR,
  //           name: 'red',
  //           value: '#ff0000',
  //         },
  //       },
  //     },
  //   ], {
  //     commitMessage: '',
  //   });

  //   expect(mockCreateOrUpdateFiles).toBeCalledWith({
  //     branch: 'main',
  //     owner: 'MattOliver',
  //     repo: 'figma-tokens-testing',
  //     createBranch: false,
  //     changes: [
  //       {
  //         message: 'Initial commit',
  //         files: {
  //           'data/tokens.json': JSON.stringify(
  //             {
  //               $themes: [
  //                 {
  //                   id: 'light',
  //                   name: 'Light',
  //                   selectedTokenSets: {
  //                     global: TokenSetStatus.ENABLED,
  //                   },
  //                 },
  //               ],
  //               global: {
  //                 red: {
  //                   type: TokenTypes.COLOR,
  //                   name: 'red',
  //                   value: '#ff0000',
  //                 },
  //               },
  //             },
  //             null,
  //             2,
  //           ),
  //         },
  //       },
  //     ],
  //   });
  // });
```

**Review by six7 — CHANGES_REQUESTED** [review]

**Comment by six7** [comment]
@TheeMattOliver I merged next in, fixed the merge conflicts and fixed some of your tests. You had a test that was testing that we can write, so I commented that out as it doesnt work yet. If possible apply those suggestions then we can merge.

File: src/storage/BitbucketTokenStorage.ts (hunk 3 above)
Comment (six7) [reply src/storage/BitbucketTokenStorage.ts:63]: need to do this as otherwise we get a typescript error because you have unused variables

File: src/storage/BitbucketTokenStorage.ts (hunk 4 above)
Comment (six7) [reply src/storage/BitbucketTokenStorage.ts:111]: need to do this as otherwise we get a typescript error because you have unused variables

File: src/storage/__tests__/BitbucketTokenStorage.test.ts (hunk 7 above)
Comment (six7) [reply src/storage/__tests__/BitbucketTokenStorage.test.ts:51]: seems like that data had the wrong structure

File: src/storage/__tests__/BitbucketTokenStorage.test.ts (hunk 8 above)
Comment (six7) [reply src/storage/__tests__/BitbucketTokenStorage.test.ts:174]: i had to comment that whole test out because it's failing

**Comment by TheeMattOliver** [comment]
> @TheeMattOliver I merged next in, fixed the merge conflicts and fixed some of your tests. You had a test that was testing that we can write, so I commented that out as it doesnt work yet. If possible apply those suggestions then we can merge.

Thank you! 

Committed those changes -- let me know if there's more to do!

**Review by six7 — APPROVED** [review]
Had another conflict, fixed it! Let's merge this in so we can continue 🎉
